include/boost/lockfree/spsc_queue.hpp

```diff
@@ -382,8 +421,30 @@ class ringbuffer_base
         return write_index == read_index;
     }
 
+#ifndef BOOST_NO_CXX11_RVALUE_REFERENCES
+    template< class OutputIterator >
+    typename boost::enable_if< typename boost::is_constructible<T, T>::type, OutputIterator>::type
+    move_and_delete( T * first, T * last, OutputIterator out )
+    {
+        if (boost::has_trivial_destructor<T>::value) {
+            return std::move(first, last, out); // will use memcpy if possible
```

Comment:
Instead of duplicating this function, and using SFINAE to select the desired one, I would just have one implementation. This is safe because while, during overload resolution, the compiler will prefer move assignment, it will still fall back to copy assignment if no move assignment operator is available.

Additionally I'd use [`boost::move`](http://www.boost.org/doc/libs/1_65_1/doc/html/move/move_algorithms.html) from `<boost/move/algorithm.hpp>` and `boost::move` from `<boost/move/utility_core.hpp>`. That will work without having to use the preprocessor (it's valid regardless of whether `BOOST_NO_CXX11_RVALUE_REFERENCES` is defined) and additionally supports types that implement move emulation too.